src/common/pico_sync/include/pico/sem.h

```diff
@@ -122,6 +122,16 @@ bool sem_acquire_timeout_us(semaphore_t *sem, uint32_t timeout_us);
  */
 bool sem_acquire_block_until(semaphore_t *sem, absolute_time_t until);
 
+/*! \brief Attempt to acquire a permit from a semaphore without blocking
```

Comment:
Apologies if this is a stupid question (it's been many many years since I wrote any semaphore-related code), but is this basically the same as calling `sem_acquire_timeout_us` but with a 0 timeout value?

Reply:
If you call `sem_acquire_block_until()` with the timeout set to expire immediately, and there are no permits available, then it will call `lock_internal_spin_unlock_with_best_effort_wait_or_timeout()` with the immediate expiration time and return false. That amounts to nearly the same thing, except that the timeout has to be evaluated. `lock_internal_etc_etc` is a macro that expands to call `spin_unlock()`, as we have here, and then `best_effort_wfe_or_timeout(until)`.

So you're right that it's very similar, I hadn't noticed that. But `sem_try_acquire()` saves a few cycles, since there's no need to set an "immediate timeout" value that is then evaluated as expiring right away.

Reply:
> But `sem_try_acquire()` saves a few cycles, since there's no need to set an "immediate timeout" value that is then evaluated as expiring right away.

Indeed, and that's always a good thing :+1: